src/gui/attributetable/qgsattributetablemodel.cpp

```diff
@@ -235,7 +236,13 @@ void QgsAttributeTableModel::featureAdded( QgsFeatureId fid )
         QgsFieldFormatter *fieldFormatter = mFieldFormatters.at( cache.sortFieldIndex );
         const QVariant &widgetCache = mAttributeWidgetCaches.at( cache.sortFieldIndex );
         const QVariantMap &widgetConfig = mWidgetConfigs.at( cache.sortFieldIndex );
-        const QVariant sortValue = fieldFormatter->representValue( mLayer, cache.sortFieldIndex, widgetConfig, widgetCache, mFeat.attribute( cache.sortFieldIndex ) );
+        // If using the default formatter use the raw value for sorting
+        // (keep the correct QVariant type and do not convert to a possibly localized string)
+        // See: https://github.com/qgis/QGIS/issues/34935
+        const bool isFallbackFormatter { fieldFormatter->id().isEmpty() };
+        const QVariant sortValue = isFallbackFormatter ?
+                                   mFeat.attribute( cache.sortFieldIndex ) :
+                                   fieldFormatter->representValue( mLayer, cache.sortFieldIndex, widgetConfig, widgetCache, mFeat.attribute( cache.sortFieldIndex ) );
```

Comment:
I think this should be using fieldFormatter->sortValue( ... ) instead. That already has appropriate logic in the field formatter subclasses to correctly handle sort order (and is used elsewhere in this class)